File: apps/server/src/routes/kanka.ts
import type { Hono } from 'hono';
import { config } from '../config.js';

const KANKA_API_BASE = 'https://api.kanka.io/1.0';

interface CacheEntry {
  date: number; // epoch millis
  data: Buffer;
  contentType: string;
}

const cache = new Map<string, CacheEntry>();

/**
 * Mounts the caching Kanka proxy at `/kanka/*`, mirroring the original Ktor
 * implementation (bearer auth, per-path cache with a configurable delay, and
 * the X-Cache-* response headers).
 */
export function registerKanka(app: Hono): void {
  const { apiKey, cacheDelaySeconds } = config.kanka;

  app.get('/kanka/*', async (c) => {
    if (apiKey == null) {
      return c.text('Kanka integration is disabled', 503);
    }

    const path = new URL(c.req.url).pathname.replace(/^\/kanka/, '');
    const apiPath = `${KANKA_API_BASE}${path}`;

    c.header('X-Cache-Delay', String(cacheDelaySeconds));

    const cached = cache.get(apiPath);
    const fresh =
      cached != null &&
      cached.date + cacheDelaySeconds * 1000 >= Date.now();
    if (cached != null && fresh) {
      c.header('X-Cache-Hit', 'true');
      c.header('X-Cache-Date', new Date(cached.date).toISOString());
      return c.body(new Uint8Array(cached.data), 200, {
        'Content-Type': cached.contentType,
      });
    }

    const apiResponse = await fetch(apiPath, {
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${apiKey}`,
      },
    });

    const bytes = Buffer.from(await apiResponse.arrayBuffer());
    const contentType =
      apiResponse.headers.get('content-type') ?? 'application/json';

    if (apiResponse.ok) {
      cache.set(apiPath, { date: Date.now(), data: bytes, contentType });
    } else {
      cache.delete(apiPath);
    }

    c.header('X-Cache-Hit', 'false');
    apiResponse.headers.forEach((value, key) => {
      if (key.toLowerCase().startsWith('x-')) {
        c.header(key, value);
      }
    });

    return c.body(new Uint8Array(bytes), apiResponse.status as 200, {
      'Content-Type': contentType,
    });
  });
}

File: legacy/src/main/kotlin/com/lapis/config/Kanka.kt
package com.lapis.config

import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.logging.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.server.application.*
import io.ktor.server.request.*
import io.ktor.server.response.*
import io.ktor.server.routing.*
import io.ktor.util.*
import kotlinx.datetime.Clock
import kotlinx.datetime.DateTimeUnit
import kotlinx.datetime.Instant
import kotlinx.datetime.plus

const val KANKA_API_BASE = "https://api.kanka.io/1.0"

fun Application.setupKankaRouting()
{
	val apiKey = environment.config.propertyOrNull("kanka.api_key")?.getString()
	val cacheDelay = environment.config.propertyOrNull("kanka.cache_delay")?.getString()?.toLongOrNull() ?: 60L
	
	val kankaClient = HttpClient(CIO) {
		install(Logging)
		expectSuccess = true
	}
	
	val cache = mutableMapOf<String, CacheEntry>()
	
	routing {
		get("/kanka/{...}") {
			if (apiKey == null) {
				call.respondText("Kanka integration is disabled", status = HttpStatusCode.ServiceUnavailable)
				return@get
			}
			
			val path = call.request.uri.removePrefix("/kanka")
			val apiPath = "${KANKA_API_BASE}$path"
			
			call.response.headers.append("X-Cache-Delay", cacheDelay.toString())
			
			val cachedEntry = cache[apiPath]
			if (!(cachedEntry == null || cachedEntry.date.plus(cacheDelay, DateTimeUnit.SECOND) < Clock.System.now())) {
				call.response.headers.append("X-Cache-Hit", "true")
				call.response.headers.append("X-Cache-Date", cachedEntry.date.toString())
				call.respondBytes(cachedEntry.data, contentType = ContentType.Application.Json)
				return@get
			}
			
			val apiResponse = kankaClient.get(apiPath) {
				headers {
					contentType(ContentType.Application.Json)
					bearerAuth(apiKey)
				}
			}
			
			val apiBytes = apiResponse.readRawBytes()
			
			if (apiResponse.status.isSuccess()) {
				cache[apiPath] = CacheEntry(
					Clock.System.now(),
					apiBytes
				)
			}
			else {
				cache.remove(apiPath)
			}
			
			call.response.headers.apply {
				append("X-Cache-Hit", "false")
				apiResponse.headers
					.filter { key, _ -> key.startsWith("X-") }
					.forEach { key, values ->
						values.forEach { value ->
							append(key, value)
						}
					}
			}
			call.respondBytes(apiBytes, contentType = apiResponse.contentType(), status = apiResponse.status)
		}
	}
	
	if (apiKey == null) {
		kankaClient.close()
	}
}

data class CacheEntry(val date: Instant, val data: ByteArray)
{
	override fun equals(other: Any?): Boolean
	{
		if (this === other) return true
		if (javaClass != other?.javaClass) return false
		
		other as CacheEntry
		
		if (date != other.date) return false
		if (!data.contentEquals(other.data)) return false
		
		return true
	}
	
	override fun hashCode(): Int
	{
		var result = date.hashCode()
		result = 31 * result + data.contentHashCode()
		return result
	}
}